### ISSUE TITLE
rez-release allows overwriting already released version on commit that isn't the release tag

### ISSUE THREAD

**Issue by mmdanggg2** [AcademySoftwareFoundation/rez#1498]
rez-release allows you to release and overwrite an already released package version while on a different commit to the one tagged at release.

**Environment**
* OS CentOS Linux release 7.6.1810 (Core) x86_64
* Rez version 2.112.0
* Rez python version Python 3.6.8

**To Reproduce**
1. Create a simple rez package with a git repo and remote
In this example my package.py is just this:
```
name = "test_pkg"
version = "1.0.0"
build_command = False

def commands():
    print("This is version 1.0.0!")
```
2. git commit, push and then rez-release the package.
3. Verify the release completed by checking in `git log` that the commit has been tagged as `test_pkg-1.0.0` and the package is ok with `rez-env test_pkg==1.0.0` and see the text is printed.
4. Modify the package in some way and leave the version as 1.0.0.
In this case, update the statement in the package.py to
```
def commands():
    print("This is version 1.0.0! BUT MODIFIED!")
```
6. git commit, push this change.
7. Try to rez-release the package again and see that it now releases version 1.0.0 again, overwriting the originally released version!
9. Verify the now incorrectly released package with `rez-env test_pkg==1.0.0` and see it prints the modified message and the `git log` still shows the previous commit as the tagged release

**Expected behavior**
I expect there to be an error thrown when trying to release a version of a package that has already been released while I am not on the commit that has been tagged for that version.

**Actual behavior**
The release is allowed to go through dispite the package now not matching with the commit that is tagged for that version.

**Comment by JeanChristopheMorinPerso** [comment]
@mmdanggg2 when you say the released package is overwritten, you mean it's completely overwritten? Because if that's the case, that's pretty bad. Also, I'd be curious to know what your build looks like (just to make sure we can fully repro).

**Comment by mmdanggg2** [comment]
Yes, I do mean that the files within the release_packages_path for that version are overwritten. In that simple example package.py I put above, the package.py and build.rxt files are overwritten. I suppose it depends on your build script how much gets clobbered. Rez does not clear out the folder it installs to first so any files that were installed previously will remain but they will be overwritten if the build script goes and copies them from the source again.

I've not been using the cmake build system because I'm not very familiar with it so my builds are just calling a python script which, for most stuff, is just copying files from source to install with shutil stuff.

**Comment by herronelou** [comment]
I was just able to reproduce this, I had a minor change on a package and though I'd try to release again without changing the version number to see if I could repro this, and it did re-release it without complaints.

**Comment by JeanChristopheMorinPerso** [comment]
If someone is interested in helping us on this, it would be appreciated. I'm concerned about this but I unfortunately won't have time to take a look at it for a couple of weeks.

Types of help that could be done: root cause analysis, code inspection, hypothesis as to why it's happening, a PR, pointing to the problematic code, proposing a fix with a diff provider here (if the CLA is a problem), etc.

**Comment by herronelou** [comment]
related to https://github.com/AcademySoftwareFoundation/rez/issues/914 (or even maybe same issue)

**Comment by herronelou** [comment]
I dug into this and have some update information. @JeanChristopheMorinPerso you mentioned earlier any form of help would be appreciated, so hope this helps you guys come up with a fix, I'm hesitant to patch it as I'm not sure about some of the existing code and if changing it might break something else.

### Updated repro / tests:

I have tested a few different cases:

#### Package without variants
- Modifying any file other than package.py : The release is properly blocked `WARNING  Skipping 1/1: destination variant already exists`
- Modifying package.py (regardless of changes in other files or not): Package can be re-released and no warning or error are raised.

#### Package with variants
- Just like the non-variant package, as long as package.py is not modified, everything is fine.
- If package.py is modified, the first variant gets overwritten, but the other variants properly get detected as existing `1 of 2 releases were successful`

### Root cause analysis

`rez-release` checks whether a variant is already released by doing a dry-run install in `LocalBuildProcess._release_variant()`:

https://github.com/AcademySoftwareFoundation/rez/blob/f19dad25ec9cbb6b0e9f53ff2ec5c964b077912a/src/rezplugins/build_process/local.py#L394-L404


The key detail is that `_create_variant(..., dry_run=True)` only reports an existing installed variant if the package definition has **not** changed:

https://github.com/AcademySoftwareFoundation/rez/blob/6654418abea2fa0de7381dae2aff1b2e004811ce/src/rezplugins/package_repository/filesystem.py#L1363-L1367

`package_changed` is computed by comparing package-level data between the existing released package and the package being released:

https://github.com/AcademySoftwareFoundation/rez/blob/6654418abea2fa0de7381dae2aff1b2e004811ce/src/rezplugins/package_repository/filesystem.py#L1311-L1325

This means that if I change something package-level like `description`, `commands()`, etc, the dry-run duplicate check returns `None`, and release proceeds even though the package version already exists.

I'm not sure why the code was made to return None if the package was modified, maybe this should check if the `version` was modified? The code was introduced in https://github.com/AcademySoftwareFoundation/rez/commit/fa1a91c8cc01bf50f158ec740e5bcba18f27a9dd#diff-0923f674fb98267b2cbc6111a14e4073edceab2b1b2661939dfef8bf7f112dbc alongside a good number of other changes.

This explains the behavior, as it's doing exactly what the code implemented.
I assume that for the multi-variant case, the first variant overwrites the package.py, after which the other variants see `package_changed` as False, as it's comparing against the freshly installed one.

### Conclusion

This looks like the duplicate detection used by `rez-release` is conflating:

- “an equivalent variant already exists”
- with
- “the package definition has not changed”

For release safety, those probably should not be treated as the same thing. In particular, if the same package name/version already exists and the package definition has changed, that seems like exactly the case that should error out rather than silently re-release/overwrite.

**Comment by herronelou** [comment]
actually commit https://github.com/AcademySoftwareFoundation/rez/commit/e993216161eef75fc598f90e3819477121c529ca#diff-0923f674fb98267b2cbc6111a14e4073edceab2b1b2661939dfef8bf7f112dbc is where this was introduced, it mentions PR https://github.com/AcademySoftwareFoundation/rez/pull/163, the previous commit I had above was a refactoring

**Comment by maxnbk** [comment]
Regarding the general topic, IMHO, this isn't actually something to solve in rez.
(Not speaking about inconsistencies in rez judgement of what is a package.py with changed def, etc)

This is normally solved with IT/systems controls.

For linux, you use a umask on a top-level package-repository. Files that are written get permissions implicitly stripped down to the imposed limit and become unwriteable-on-top-of, or you use release-hooks to drop-permissions after writing them, or change a group permission to a more isolated one.
On windows, you use inheritance of folder-level permissions for new folders created within that one.

Other setups are possible with CI using certain network accounts, group-permissions, and similar types of workflows, depending on a studios predisposition, file systems, network protocols, and so on.

I'm not against us investing in handholds/guardrails/guidance for this, but it truly isn't a one-size-fit-all space, and _traditionally_ this is the realm of IT and what they do, but we're all just too busy cramming "all the things we want in one tool, into one tool", when delegation of responsibility is needed. This is one such case.

(I am still grateful for @herronelou digging in on the discrepancy bits, and I haven't reviewed that yet, I just wanted to drop this because I read the topic while I was on a plane and wanted to not forgot to write it up)

**Comment by herronelou** [comment]
@maxnbk I do believe this is a bug, Rez attempts to prevent re-releasing a package, but right now it fails to in certain cases (if any changes were made to package.py).
This is not a documented behavior, nor can I imagine this would be the intended one, it's pretty inconsistent.
I do believe rez should have a basic existence check like it currently does (if we wanted to make that optional that'd be fine, but I wouldn't remove it), but it needs to be fixed.

**Comment by maxnbk** [comment]
Yes, I understand. I am just asserting that "locking down packages permissions-wise" is a studio/network/IT concern, not that rez should attempt to overwrite an existing release.
(I could however see people wanting to configure that behavior)
